4단계 - EntityEntry

catsbi opened the pull request:
안녕하세요 리뷰어님 어느덧 4단계 미션까지 왔네요 ㅎㅎ

진행하면서 후반부에 id가 없는 신규 엔티티를 어디서 아이디를 생성해줘야하는지 고민하다가 처음엔 DefaultEntityManager에서 EntityPersister를 이용해 직접 생성하다가 해당 계층보다는 EntityEntry가 상태를 관리하는 객체로 Saving 상태일 경우이면서 id가 없을 경우 아이디 생성전략에 따라 직접 생성 후 EntityEntry를 생성하도록 해줬습니다. 🤔 

그럼 이번 미션도 잘 부탁드려요!

lhs0610 committed:
feat: EntityEntry 및 Status 생성

lhs0610 committed:
feat: 엔티티 영속성 상태 관리 객체(EntityEntry)로 변경

lhs0610 committed:
fix: 테스트 케이스 변경된 스펙에 맞게 수정

lhs0610 committed:
fix: dirty check 이후 스냅샷 업데이트 로직 추가 및 테스트 케이스 작성

lhs0610 committed:
test: EntityEntry 테스트 케이스 추가

- 상태에 따른 더티체킹 테스트 케이스 추가

lhs0610 committed:
test: 식별자 필드 GeneratedValue 애노테이션 체크 후 분기로직 추가

jongmin4943 commented on src/main/java/persistence/sql/dml/impl/DefaultEntityManager.java:
```diff
@@ -46,9 +48,10 @@ public <T> void persist(T entity) {
         }
 
-        Object id = entityPersister.insert(entity);
-        persistenceContext.add(id, entity);
-        persistenceContext.createDatabaseSnapshot(id, entity);
+        EntityEntry entityEntry = persistenceContext.addEntry(entity, Status.SAVING, entityPersister);
+        if (!transaction.isActive()) {
+            entityEntry.dirtyCheck();
+        }
```
`persist` 시작 시 SAVING 상태를 잘 만들어 주셨네요 💯 
그런데, `persist` 에 transaction 이 inactive 상태일때 dirtyCheck 가 일어나는건 어색하다고 느껴져요 🤔

catsbi replied:
transaction이 begin상태가 아닐 경우 요청마다 바로 변경사항 적용을 하라는 의미였는데, 그럴 필욘 없을까요 🤔

jongmin4943 replied:
조건문은 괜찮은데 `dirtyCheck` 로 하는게 어색하게 느껴졌어요 :)

jongmin4943 replied:
조건문보단 `dirtyCheck` 로 행위가 일어나는게 어색하다 느꼈어요 :)

jongmin4943 commented on src/main/java/persistence/sql/dml/impl/DefaultEntityManager.java:
```diff
@@ -98,8 +102,11 @@ public <T> void remove(T entity) {
+        entityEntry.updateStatus(Status.DELETED);
+        if (!transaction.isActive()) {
+            entityEntry.dirtyCheck();
+        }
```
마찬가지로 `remove` 시 에도 `dirtyCheck` 가 필요한 이유가 있을까요? 

Hibernate 에서는 `dirtyCheck` 는 변경감지 로직으로, 주로 업데이트와 관련된 작업에 사용되고 있어요. flush 시점에 한번에 반영하기 위한 로직처럼 느껴지는데, `dirtyCheck` 가 아닌 각각 순회하며 상태에 맞는 처리를 하는게 어떨까요?

jongmin4943 commented on src/main/java/persistence/sql/dml/impl/DefaultEntityManager.java:
```diff
@@ -108,18 +115,17 @@ public <T> T find(Class<T> returnType, Object primaryKey) {
 
         T loadedEntity = entityLoader.load(primaryKey);
         if (loadedEntity != null) {
-            persistenceContext.add(primaryKey, loadedEntity);
-            persistenceContext.createDatabaseSnapshot(primaryKey, loadedEntity);
+            persistenceContext.addEntry(loadedEntity, Status.MANAGED, entityPersister);
```
`Status.LOADING` 이 사용되지 않았네요. 😢 
LOADING 은 언제 사용되어야 할까요?

catsbi replied:
ID와 returnType을 기반으로 LOADING중인 EntityEntry를 만들어줬습니다. 

그럼 findAll같은 경우에는 primaryKey가 조회전에는 존재하지 않는데 LOADING중인 EntityEntry를 만들 수 없으려나요?

jongmin4943 replied:
현재 요구사항으로는 사실 명확한 차이를 만들기 어렵다고 느끼긴 합니다.
그래도 만들어본다면 `EntityEntry` 를 key 기반으로 먼저 `LOADING` 으로 등록하고, load 한 뒤에 해당 key 의 `EntityEntry` 의 상태를 `MANAGED` 로 변경하는게 미션의 방향이었습니다 👍

jongmin4943 replied:
> 그럼 findAll같은 경우에는 primaryKey가 조회전에는 존재하지 않는데 LOADING중인 EntityEntry를 만들 수 없으려나요?

사실 `EntityManager` spec 에는 `findAll` 이 없기때문에 너무 깊게 고려 안하셨으면 해요 👍

jongmin4943 commented on src/main/java/persistence/sql/entity/EntityEntry.java:
```diff
@@ -0,0 +1,175 @@
+import java.lang.reflect.Field;
+import java.util.List;
+
+public class EntityEntry {
+    private final MetadataLoader<?> loader;
+    private final EntityPersister entityPersister;
+    private final PersistenceContext persistenceContext;
```
개인적으로 `EntityEntry` 가 `EntityPersister` 를 가지고 직접 데이터베이스와 작업하는 책임까지 가지기에는 조금 무겁다고 느껴져요 😢 `EntityEntry` 는 `영속성 컨텍스트 내에서 Entity 의 상태를 관리` 하는 책임을 위주로 가져보면 좋을 것 같아요.

> EntityEntry 클래스는 엔터티의 영속성 상태와 상태 변화, 생명주기와 변경감지에 중요한 역할

또한 `EntityEntry` 와 `PersistenceContext` 가 양방향으로 순환참조 되고 있고 있네요. 저는 양방향 참조를 가능하면 지양하고 있어요.
- 객체 간 관계가 복잡해지게 된다면 코드의 가독성이 떨어지고, 코드 흐름을 파악하기 어려워 진다 느껴요.
- 테스트 작성이 어렵고, 문제가 발생했을 때 디버깅도 까다로워질 수 있어요.
- 두 객체 간의 결합도가 매우 높아지고 응집도가 떨어지기 쉽다고 느껴요.

한솔님은 어떻게 생각하시나요? :)

catsbi replied:
AbstractEntityEntry는 상태로 EntityPersister를 받아서 (비록 메타데이터 조회뿐이지만) 사용을 하고 있기에 의존성이 생긴이상 함수를 사용해도 더 큰 문제는 없지 않을까? 라는 생각을 했습니다. 

그리고 그 와중 Visitor Pattern을 활용한다는 개념으로 문제가 없다고 판단했어요. EntityEntry의 정의를 보면 db통신은 persister로 옮기는게 맞을 것 같네요. 그리고 양방향 참조에 대해서도 함부로 사용해선 안되지만, 옵저버패턴도 그렇고, 방문자패턴도 그렇고, 사용하기에 따라 유용하게 사용할 수 있을 것 같아 학습과정인것을 감안해 시도해봤습니다 ㅎ

catsbi replied:
그런데 EntityEntry에서 EntityPersister의 db 통신(쿼리수행)을 제거하려고하니 insert 쿼리수행같은 경우는 
SAVING상태로 저장을 하려하더라도 id가 없기때문에 엔트리 생성시점에 insert쿼리가 수행되어야 할 것 같습니다. 

그런데 entityManager에서 insert쿼리를 수행 한 다음 엔트리 추가를 해주면 이미 saving이 아니라 managed상태인것 같아요. 
이 경우 어떻게 해야할까요?

jongmin4943 replied:
이 경우에도 `LOADING` 과 비슷한 방향이에요. key 기반으로 먼저 `SAVING` 으로 등록하고 실제 insert 가 일어날때 `MANAGED` 로 바꿔주는거죠. 여기서 `SAVING` 시 key 가 필요하기 때문에 고려해주신 것 처럼 id 생성 전략에 따라 key 를 먼저 채번하고, 바로 `MANAGED` 가 될 수 도 있겠죠 :)

jongmin4943 commented on src/main/java/persistence/sql/entity/EntityEntry.java:
```diff
@@ -0,0 +1,175 @@
+
+    }
+
+    private static void createEntityIdOrThrow(EntityPersister entityPersister, Object entity, GeneratedValue anno) {
+        // TODO 현재는 IDENTIY만 지원한다.
+        if (anno != null && anno.strategy() == GenerationType.IDENTITY) {
+            entityPersister.insert(entity);
+            return;
+        }
```
멋진 고려사항이네요 👍

jongmin4943 reviewed (changes requested):
안녕하세요 한솔님! 

마지막인 4단계 미션 진행해주셨네요 👍 
몇 가지 궁금한 점에 대해 코멘트 달아보았습니다. 확인 후 재요청 부탁드릴게요. 🙇

궁금한 사항은 언제든 추가 코멘트 또는 DM 주세요! 😃

lhs0610 committed:
feat: 피드백 반영

- 엔티티 조회 중간과정의 LOADING Entry 생성로직 추가

lhs0610 committed:
feat: 피드백 반영

- EntityManager 의 dirtyCheck로직 이동
- PersistenceContext 테스트 케이스 추가

lhs0610 committed:
delete: 불필요 EntityEntry 인스턴스 변수 제거

catsbi commented:
리뷰어님 피드백 및 코멘트 답글달아주신것 기반으로 리팩터링 진행해봤습니다. 

다른것보다 각 계층간의 책임의 경계가 명확하게 알고 있지를 않다보니, 이리저리 헤매는 코드가 되는 것 같습니다 😓 
그래도, 하나씩 익혀가면서 만들어나가고 피드백통해서 흐릿한 부분들이 명확해지는건 좋은 것 같습니다.

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -1,163 +1,99 @@
+    private final EntityPersister persister;
 
-        return null;
+    public DefaultPersistenceContext(EntityPersister persister) {
+        this.persister = persister;
     }
```
`PersistenceContext` 가 상태로 persister 를 가지게 되었네요.
현재 만들어주신 `DefaultPersistenceContext` 는 상태로도 persister 를 가지기도 하고, 파라미터로도 persister 가 넘어오기도 하는 것 같아요. 해당 부분에서 조금은 혼란이 있을 것 같아 정리를 하면 좋을 것 같아요.

![image](https://github.com/user-attachments/assets/4d74b0cb-861d-43c9-bfdc-cdc9692706db)
> Persistence Context는 엔티티 객체가 관리되는 메모리 공간을 의미. 즉, EntityManager가 관리하는 엔티티 객체들의 저장소이며, 데이터베이스와 애플리케이션 사이의 중간 저장 공간 역할

개인적으로 `PersistenceContext` 는 persister 를 가지지 않고 Entity 들을 관리하는 책임을 가지고, `EntityManager` 가 persister 를 이용하는 방향을 선호하지만, 한솔님 의견도 중요한 것 같아요. `PersistenceContext` 가 직접 persister를 가지는것에 대해 어떻게 생각하시나요?

catsbi replied:
음 제가 봤던건 그 아래에 있던 내용이였는데, PersistenceContext는 상태관리에 있어서
`Persistence Context는 엔티티를 관리하며, 영속성 상태의 엔티티를 감지하여 데이터베이스에 동기화함`
이라는 항목이 있어서 persister를 이용해서 직접 데이터베이스에 동기화하는 로직을 넣어줬습니다. 

저는 해당 정리 표를 보고 역할은 엔티티 객체를 메모리공간에서 관리하는 1차 캐싱 역할과 상태를 관리하여 변경이 필요한경우 데이터베이스와 소통하여 동기화시켜주는 역할을 하도록했습니다. 그리고 엔티티 매니저에서는 상태를 신경쓰지 않아도 되는 not transactional상태일 경우 데이터베이스와 소통하도록 하는방식으로 구현했습니다.

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -1,163 +1,99 @@
-        KeyHolder key = new KeyHolder(entity.getClass(), id);
-        context.remove(key);
-        snapshot.remove(key);
+    public <T> EntityEntry addEntry(Object primaryKey, Class<T> returnType, Status status, EntityPersister entityPersister) {
+        EntityEntry entityEntry = EntityEntry.newLoadingEntry(primaryKey, returnType);
+        context.put(entityEntry.getKey(), entityEntry);
+
+        return entityEntry;
     }
```
이 메서드에서 `Status status, EntityPersister entityPersister` 이 두가지 파라미터는 안쓰이는 것 같은데 필요한 이유가 있을까요?

catsbi replied:
후반부 스펙이 계속바뀌면서 지우는걸 놓친것 같네요.

jongmin4943 commented on src/main/java/persistence/sql/dml/impl/DefaultEntityManager.java:
```diff
@@ -130,16 +144,16 @@ public <T> List<T> findAll(Class<T> entityClass) {
-        if (persistenceContext.isDirty()) {
-            persistenceContext.getDirtyEntities().forEach(this::merge);
-        }
+        persistenceContext.dirtyCheck();
```
아무래도 flush 시 `dirtyCheck` 를 통해 db에 반영하는 로직 때문에 `DefaultPersistenceContext` 가 persister 를 상태로 가지게 된 것 같아요. 몇 가지 방향이 있을 것 같아요.
- persister 를 이용한 db 반영은 `EntityManager` 가 담당한다.
- `dirtyCheck` 파라미터로 persister 를 넘긴다.

한솔님은 어떻게 생각하시나요? :)

catsbi replied:
entityPersister를 넘겨서 처리하도록 진행해봤습니다. 위에서 코멘트단것처럼 db와 소통하며 동기화하는건 persistenceContext가 맞는거 같아서요!

jongmin4943 commented on src/main/java/persistence/sql/entity/EntityEntry.java:
```diff
@@ -0,0 +1,170 @@
+                key);
+    }
+
+    public static EntityEntry newEntry(EntityPersister entityPersister, PersistenceContext persistenceContext, Object entity, Status status) {
```
이 메서드에서 파라미터인 `persistenceContext` 는 쓰이지 않고 있네요.

아직 `persister` 에 대한 의존성이 아직 남아있는데, Id 전략에 따른 저장로직이 있어서 그런것 같아요.
개인적으로 해당 부분도 `EntityEntry` 보단 `EntityManager` 가 하면 어떤가 싶은데 한솔님 의견이 궁금합니다 :)

catsbi replied:
EntityManager로 바꿔줬습니다. 그러다보니 Saving상태를 고려해서 그럼 언제 SAVING이고 언제 MANAGED인지 생각했는데
Entitymanager에서 트랜잭션 유무에따라 MANGED상태 업데이트 로직을 추가하는 방향으로 풀었습니다.

jongmin4943 reviewed (changes requested):
안녕하세요 한솔님! 

4단계 미션 피드백 반영해주셨네요 👍 
미션을 마무리 하기 전에 한번 더 확인해보면 좋을 만한 부분에 코멘트 달아보았어요. 확인 후 재요청 부탁드릴게요. 🙇

궁금한 사항은 언제든 추가 코멘트 또는 DM 주세요! 😃

lhs0610 committed:
refactor: 피드백 반영

- 불필요 테스트 케이스 제거
- PersistenceContext의 entityPersister 상태 제거 및 매개변수로 받도록 수정
- persist 데이터베이스 통신 로직 EntityManager로 이동

lhs0610 committed:
test: EntityEntry 테스트 케이스 추가

lhs0610 committed:
test: EntityEntry 테스트 케이스 추가

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -12,22 +12,16 @@
-    }
-
     @Override
     public <T> EntityEntry addEntry(T entity, Status status, EntityPersister entityPersister) {
```
```suggestion
    public <T> EntityEntry addEntry(T entity, Status status) {
```
여기있는 `EntityPersister` 도 현재는 필요 없어 보이네요 :)

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -52,41 +46,41 @@ public <T, ID> void deleteEntry(T entity, ID id) {
-            handleEntry(entry);
+            handleEntry(persister, entry);
         }
     }
```
> entityPersister를 넘겨서 처리하도록 진행해봤습니다. 위에서 코멘트단것처럼 db와 소통하며 동기화하는건 persistenceContext가 맞는거 같아서요!

넘겨주시는 방향으로 선택해주셨네요. 👍 저도 현 구조가 좀 더 깔끔하게 보여요 :)

jongmin4943 commented on src/main/java/persistence/sql/dml/impl/DefaultEntityManager.java:
```diff
@@ -47,8 +47,11 @@ public <T> void persist(T entity) {
         }
-
-        persistenceContext.addEntry(entity, Status.SAVING, entityPersister);
+        entityPersister.insert(entity);
+        EntityEntry entityEntry = persistenceContext.addEntry(entity, Status.SAVING, entityPersister);
+        if (!transaction.isActive()) {
+            entityEntry.updateStatus(Status.MANAGED);
+        }
```
> EntityManager로 바꿔줬습니다. 그러다보니 Saving상태를 고려해서 그럼 언제 SAVING이고 언제 MANAGED인지 생각했는데
Entitymanager에서 트랜잭션 유무에따라 MANGED상태 업데이트 로직을 추가하는 방향으로 풀었습니다.

피드백에 대해 고민 해주시고 반영해주셨네요. 감사합니다 🙇

jongmin4943 commented on src/main/java/persistence/sql/config/PersistenceConfig.java:
```diff
@@ -69,7 +69,7 @@ public EntityPersister entityPersister() throws SQLException {
     }
 
     public PersistenceContext persistenceContext() throws SQLException {
```
```suggestion
    public PersistenceContext persistenceContext() {
```
사소하지만, throws 구문은 없어져도 될 것 같네요 👍

jongmin4943 reviewed (approved):
안녕하세요 한솔님! 

4단계 미션 피드백 반영 잘 해주셨네요 👍
바쁘신 와중에도 쉽지 않으셨을텐데 피드백에 대해 고민해주시고 받아주셔서 감사합니다 🙇 도전하시는 모습에 저도 많은 관점을 얻을 수 있었어요! 😃

미션 동안 많은 것을 얻어가셨으면 좋겠습니다.
남은 미션 완주 응원하겠습니다! 🚀
